### DISCUSSION TITLE
Running Dagster in a highly data-sensitive environment?

### DISCUSSION THREAD

**Discussion by zhoekstraTMNL** [dagster-io/dagster#15608] [General]
I'm currently in my first year of working for a company that does big-data in a highly government regulated industry (hosted in AWS). Our current ETL orchestration is... not great (step functions and glue jobs), and I'd like to investigate and propose moving to dagster. Problem is, there's a few hard requirements I don't know if dagster can handle, and I'd appreciate some guidance from community experts to see if it's even worth investigating dagster as a solution.

1) Complete code and data control. This makes using dagster cloud out of the question - we need all orchestration and execution to happen inside our AWS accounts. I'm looking at some of the terraform projects that can deploy a full inststance to AWS, but if anyone has any suggestions let me know.

2) Fine-grained per-process access to resources. We need every process or job to explicitly define what roles and permissions they can use to access or operate on resources. Ex: if we have an op that reads s3 files from a bucket and writes to a table, it needs to use an AWS role that can only access those exact files and only that table. Ideally every op/asset gets a seperate role rather than a role per job/materialization flow. We can define these in terraform, but how to get those roles individually to each op while giving the orchestrator itself as little access as possible is a tough problem.

3) Access controls, or at least action audit logging. We need to know who triggered what for what reason. Ideally we have a way to make sure only people in the company with the right access can trigger specific runs. (I know cloud has Azure login support, I don't think the self-hosted solution has that)

In general, if there are any other security features for dagster that I can mention to our risk team, tell me, because while there's a lot of interest from our team, actually adopting this will largely be assuring risk and legal that we can use an orchestrator without risking a data breach

Thanks,
zhoekstra

**Comment by slopp** [comment]
Hey @zhoekstra thanks for opening this discussion!

There are a few flavors of Dagster production deployments that teams are using to meet stringent security and data privacy requirements:

1. Fully OSS. While it is possible to run the Dagster OSS stack, the default OSS stack does not include any identity or user management. Teams with strict security requirements typically prefer to use Dagster Cloud (see below) - but you can run a web proxy on top the OSS stack to restrict access. 

2. Dagster Cloud Hybrid. In this model, Dagster Cloud runs the control plane, but the orchestration code, all connections to your systems (eg warehouses), and all data access remain in your environment. Dagster Cloud has _no_ access to the data. The architecture for this model is available here: https://docs.dagster.io/dagster-cloud/deployment/hybrid. This model has a few great benefits:
-  It is SOC Type II compliant and is used by many Fortune 500 companies including those subject to GDPR or HIPPA data privacy requirements requirements. More information is available at dagster.io/security. 
- Dagster Cloud includes full identity management incl SSO, SCIM provisioning, RBAC, audit logging, and many other InfoSec requirements

To address your specific questions:

1. _Complete code and data control._ In the Dagster Cloud Hybrid model, no code or data is sent to Dagster Cloud. The only information shared by Dagster Cloud is structured metadata, as documented in the Hybrid Architecture diagram and dagster.io/security

2. _Fine-grained per-process access to resources_. I agree that this is a tough problem, and the solution will depend a bit on which job executor you are using. For Kubernetes (eg EKS), I think you could accomplish what you want through op and job tags, see: 
  - https://docs.dagster.io/deployment/guides/kubernetes/customizing-your-deployment#per-job-or-per-op-kubernetes-configuration   
  -  https://github.com/dagster-io/dagster/discussions/14544.
  
4. _Access controls_ are fully supported in Dagster Cloud. Users are required to authenticate and will have a role (Admin, Editor, Launcher, or Viewer). These roles can be applied globally or on a project-by-project basis. SSO, SCIM for provisioning users from an IdP,  and an audit log of organization events (like login attempts) are also available.

Our team is always happy to chat through these, you can get in touch via dagster.io/contact
